server/chat-plugins/abuse-monitor.ts

```diff
@@ -0,0 +1,555 @@
+	minScore: 0.65,
+	specials: {
+		THREAT: {0.96: 'MAXIMUM'},
+		IDENTITY_ATTACK: {0.8: 2},
+		SEVERE_TOXICITY: {0.8: 2},
+	},
+};
+
+export const settings: FilterSettings = (() => {
+	try {
+		// accounting for data changes -
+		// make sure we do have the default data in case it's not in the stored data
+		return {...defaults, ...JSON.parse(FS('config/chat-plugins/nf.json').readSync())};
+	} catch (e: any) {
+		if (e.code !== "ENOENT") throw e;
+		return defaults;
+	}
+})();
+
+interface FilterSettings {
+	disabled?: boolean;
+	threshold: number;
+	minScore: number;
+	specials: {[k: string]: {[k: number]: number | "MAXIMUM"}};
+}
+
+export interface PerspectiveRequest {
+	languages: string[];
+	requestedAttributes: AnyObject;
+	comment: {text: string};
+}
+
+export async function classify(text: string) {
+	const request: PerspectiveRequest = {
+		// todo - support 'es', 'it', 'pt', 'fr' - use user.language? room.settings.language...?
+		languages: ['en'],
+		requestedAttributes: ATTRIBUTES,
+		comment: {text},
+	};
+	try {
+		const raw = await Net(`https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze`).post({
+			query: {
+				key: Config.perspectiveKey,
+			},
+			body: JSON.stringify(request),
+			headers: {
+				'Content-Type': "application/json",
+			},
+		});
+		const data = JSON.parse(raw);
+		if (data.error) throw new Error(data.message);
+		const result: {[k: string]: number} = {};
+		for (const k in data.attributeScores) {
+			const score = data.attributeScores[k];
+			result[k] = score.summaryScore.value;
+		}
+		return result;
+	} catch (e) {
+		Monitor.crashlog(e, 'A Perspective API request', {request: JSON.stringify(request)});
+		return null;
+	}
+}
+
+export const PM = new ProcessManager.QueryProcessManager<any, any>(module, async ({comment}) => {
+	const now = Date.now();
+	const result = await classify(comment);
+	if (!result) return {score: 0, flags: []}; // crash. logged already.
+	const delta = Date.now() - now;
+	if (delta > 1000) {
+		Monitor.slow(`[Abuse Monitor] ${delta}ms - ${JSON.stringify({comment})}`);
+	}
+	let score = 0;
+	const flags = new Set<string>();
+	for (const type in result) {
+		const data = result[type];
+		if (data < settings.minScore) continue;
+		const curScore = score;
+		if (settings.specials[type]) {
+			for (const k in settings.specials[type]) {
+				if (data < Number(k)) continue;
+				const num = settings.specials[type][k];
+				if (num === 'MAXIMUM') {
+					score = settings.threshold;
+				} else {
+					if (num > score) score = num;
+				}
+			}
+		}
+		if (score < 1) score = 1;
+		if (score !== curScore) flags.add(type);
+	}
+	return {score, flags: [...flags]};
+}, PM_TIMEOUT, message => {
+	if (message.startsWith('SLOW\n')) {
+		Monitor.slow(message.slice(5));
+	}
+});
+
+if (!PM.isParentProcess) {
+	// This is a child process!
+	global.Config = Config;
+	global.Monitor = {
+		crashlog(error: Error, source = 'A netfilter child process', details: AnyObject | null = null) {
```

Comment:
```suggestion
		crashlog(error: Error, source = 'An abuse monitor child process', details: AnyObject | null = null) {
```

Reply:
"netfilter" makes me think of filtering network requests, not neural networks. If you want to allude to neural networks, what about nnfilter or mlfilter or neuralfilter?

Reply:
Oh, yeah. I just forgot to rename this.